### ISSUE TITLE
cqueues.notify should support watching files

### ISSUE THREAD

**Issue by daurnimator** [wahern/cqueues#80]
cqueues.notify currently only supports watching a directory.

Is there a reason that watching files isn't supported?

**Comment by daurnimator** [comment]
At least for linux, this might be as simple as removing `IN_ONLYDIR` from the `inotify_add_watch` call.

**Comment by wahern** [comment]
It does. The API is a little more abstract than inotify because the BSDs don't have anything like inotify. On BSDs you can only get change notification of a file through an open descriptor of that file. That means you can't _automatically_ track every file in a directory, because you'd have to open a descriptor to every single file.

So basically the way the notify module works is you open a directory. You then _explicitly_ add particular file names that you're interested in. For the BSDs this will attempt to open the specified file; if it doesn't exist it'll watch the directory and continually try to open the file when the directory is modified.

One of the original TODO items (i.e. issue #2) is to support glob patterns so you can easily track every file using "*" if that's what you want to do. It's just that doing it by default is a non-starter.

In some sense the BSD approach is actually preferable. By requiring a file descriptor for tracking changes, you're preallocating the necessary kernel resources needed to track files. And if the resources aren't available, you know that upfront. That means that once you open the file descriptor and install the kevent, it's _impossible_ to be able to miss events. Managing state becomes very simple in practice.

Linux inotify, by contrast, can easily overflow its internal event buffers. And it exposes some ugly sysctl knobs for limiting the amount of memory committed to inotify buffers on the system. Theoretically you can easily DoS a Linux systems by opening a ton of inotify handles and flooding the system with events. That will force constant overflowing of inotify buffers, which is signaled by a special event, IN_Q_OVERFLOW. On overflow applications are supposed to _manually_ scan every file they're tracking. But if the notify module doesn't know which files you're tracking, it's going to have to keep state on _every_ file to know how to synthesize an event, and explicitly stat every file in a directory. That could easily hammer the buffer cache and if it needs to hit the disk stall out the whole system.

Basically, this is why the notify is built around the idea of a directory. For various reasons related to both the BSD and Linux implementations, we first need to track a specific directory. Then you explicitly add filenames within the directory you're interested in.

Linux's inotify API is seemingly simple and convenient, but it hides significant ugliness. Many times in order to use it properly you basically need to do precisely what the notify module is doing, anyhow. For the times you really just want the inotify behavior, people can just use it directly.

**Comment by wahern** [comment]
Closing as duplicate of issue #2.

**Comment by daurnimator** [comment]
So perhaps there is a slight difference: I only want to be able to watch a single file, the goal is to be able to 'tail' a file.

On linux, `tail -f` opens the file and registers an inotify handle on it with `IN_MODIFY`. How does it work on BSD?

**Comment by daurnimator** [comment]
> How does it work on BSD?

Hrm, on freebsd it just stats the file once a second: https://svnweb.freebsd.org/base/head/usr.bin/tail/forward.c?view=markup#l306

How disappointing

**Comment by daurnimator** [comment]
> Hrm, on freebsd it just stats the file once a second: https://svnweb.freebsd.org/base/head/usr.bin/tail/forward.c?view=markup#l306
> 
> How disappointing

Oops, I read the comments and not the code. Apparently EVFILT_READ does work for my purposes.
This article has a bit of an intro http://doc.geoffgarside.co.uk/kqueue/file.html

**Comment by wahern** [comment]
Unlike kqueue, inotify tracks filesystem paths, not file objects. It basically attaches a watcher on a directory, and then implements some hackery to catch writes on files opened through that directory. If you do inotify_add_watch("/tmp/A/foo.txt"), if somebody moves foo.txt to "/tmp/B/foo.txt" and continues to modify it, you'll get a rename event but you won't get any more modification events until you do inotify_add_watch("/tmp/B/foo.txt"). But there's a race condition--if the file was renamed to /tmp/C/foo.txt before you could watch /tmp/B/foo.txt, you will have lost track of that file completely.

So there's no way to accurately emulate the BSD kqueue of watching modifications to a file object using inotify. But you can use BSD kqueue to emulate inotify's path-based tracking, subject to some resource constraints. And that's what the notify module does--it basically tracks paths. So if you want to watch the path /tmp/A/foo.txt, you would do

```lua
local nfy = notify.opendir"/tmp/A"
nfy:add("foo.txt")
for flags, name in nfy:changes() do
  if name == "foo.txt" do
    for flag in notify.flags(flags) --> iterate over the flags bitfield
      print(name, notify.strflag(flag))
    end
  end
end
```

This is what the examples/notify.lua do. Try it by doing:

```sh
lua ./notify.lua /tmp foo.txt
```

and in another terminal do

```sh
touch /tmp/foo.txt
echo foo >> /tmp/foo.txt
rm /tmp/foo.txt
```

For more information on the caveats with inotify, see https://lwn.net/Articles/605128/. BSD kqueue file events are super simple by comparison.

**Comment by daurnimator** [comment]
> Unlike kqueue, inotify tracks filesystem paths, not file objects. It basically attaches a watcher on a directory, and then implements some hackery to catch writes on files opened through that directory. If you do inotify_add_watch("/tmp/A/foo.txt"), if somebody moves foo.txt to "/tmp/B/foo.txt" and continues to modify it, you'll get a rename event but you won't get any more modification events until you do inotify_add_watch("/tmp/B/foo.txt"). But there's a race condition--if the file was renamed to /tmp/C/foo.txt before you could watch /tmp/B/foo.txt, you will have lost track of that file completely.

My tests with the /proc file system suggest otherwise... unless I misunderstand you?

Save as main.c:
```lua
#include <stdio.h>
#include <unistd.h>
#include <sys/inotify.h>
#include <alloca.h>
#include <limits.h>
#include <errno.h>

int main () {
	char path[20];
	int fd = inotify_init1(IN_CLOEXEC);
	if (-1 == fd) return (perror("inotify_init1"),1);
	sprintf(path, "/proc/self/fd/%d", 3);
	if (-1 == inotify_add_watch(fd, path, IN_MODIFY)) return (perror("inotify_add_watch"),1);
	struct inotify_event *returned = alloca(sizeof(struct inotify_event) + NAME_MAX + 1);
	while (1) {
		int n = read(fd, returned, sizeof(struct inotify_event) + NAME_MAX + 1);
		if (-1 == n) return (perror("read"),1);
		close(3);  // Don't care if it fails
		fprintf(stderr, "wd=%d mask=%u cookie=%u len=%u name=%s\n",
			returned->wd,
			returned->mask,
			returned->cookie,
			returned->len,
			returned->len?returned->name:0);
	}
	return 0;
}
```
```
$ gcc -Wall -Werror main.c
$ ./a.out 3< main.c &
[1] 9503
$ ls -la /proc/9503/fd
total 0
dr-x------ 2 daurnimator daurnimator  0 Jul 28 11:30 .
dr-xr-xr-x 8 daurnimator daurnimator  0 Jul 28 11:30 ..
lrwx------ 1 daurnimator daurnimator 64 Jul 28 11:30 0 -> /dev/pts/8
lrwx------ 1 daurnimator daurnimator 64 Jul 28 11:30 1 -> /dev/pts/8
lrwx------ 1 daurnimator daurnimator 64 Jul 28 11:30 2 -> /dev/pts/8
lr-x------ 1 daurnimator daurnimator 64 Jul 28 11:30 3 -> /home/daurnimator/src/inotify-test/main.c
lr-x------ 1 daurnimator daurnimator 64 Jul 28 11:30 4 -> anon_inode:inotify
```
now I just re-save main.c in my editor
```
wd=1 mask=2 cookie=0 len=0 name=(null)
wd=1 mask=2 cookie=0 len=0 name=(null)
$ ls -la /proc/9503/fd
total 0
dr-x------ 2 daurnimator daurnimator  0 Jul 28 11:30 .
dr-xr-xr-x 8 daurnimator daurnimator  0 Jul 28 11:30 ..
lrwx------ 1 daurnimator daurnimator 64 Jul 28 11:30 0 -> /dev/pts/8
lrwx------ 1 daurnimator daurnimator 64 Jul 28 11:30 1 -> /dev/pts/8
lrwx------ 1 daurnimator daurnimator 64 Jul 28 11:30 2 -> /dev/pts/8
lr-x------ 1 daurnimator daurnimator 64 Jul 28 11:30 4 -> anon_inode:inotify
```
re-save it again
```
wd=1 mask=2 cookie=0 len=0 name=(null)
wd=1 mask=2 cookie=0 len=0 name=(null)
```

**Comment by daurnimator** [comment]
> Even though the file descriptor is closed, I can still get events on it (which is interesting).
> 
> Is it following the symlink and actually watching /home/daurnimator/src/inotify-test/main.c?
> The lack of a name field suggests not?

If I rename the file and modify it at the new location, I still get events:

```
$ mv main.c main2.c 
```
(no prints)

but now I edit a single byte in main2.c:
```
wd=1 mask=2 cookie=0 len=0 name=(null)
wd=1 mask=2 cookie=0 len=0 name=(null)
```

So this seems to be working as I hoped?

**Comment by wahern** [comment]
I stand corrected. Maybe I was confusing Solaris' file events, which definitely works that way.

**Comment by daurnimator** [comment]
> Maybe I was confusing Solaris' file events, which definitely works that way.

Possibly :)
But looks like the same /proc hack works over there too: https://github.com/illumos/illumos-gate/blob/master/usr/src/cmd/tail/forward.c#L292
